File: frontend/src/services/indicators.ts
import { api } from "./api";

/** Padrão de cenário quando não informado (alinhado ao backend). */
const DEFAULT_SCENARIO_QUERY = "REALIZADO";

export interface ProjectRoi {
  project_id: string;
  project_name: string;
  competencia: string;
  scenario: string;
  revenue: number;
  cost: number;
  operational_profit: number;
  /** Fração (ex.: 0,33 = 33%). null = custo zero (ROI indefinido). */
  roi: number | null;
  roi_pct: number | null;
}

export interface RoiRanking {
  competencia: string;
  scenario: string;
  only_active: boolean;
  items: ProjectRoi[];
}

/** GET /api/v1/indicators/roi/operacional — aceita mês único (competencia) ou intervalo acumulado. */
export async function fetchRoiOperacionalRanking(params?: {
  competencia?: string;
  dataInicial?: string;
  dataFinal?: string;
  scenario?: string;
}): Promise<RoiRanking> {
  const q: Record<string, string> = { scenario: params?.scenario ?? DEFAULT_SCENARIO_QUERY };
  if (params?.dataInicial) q.data_inicial = params.dataInicial;
  if (params?.dataFinal) q.data_final = params.dataFinal;
  if (!params?.dataInicial && !params?.dataFinal && params?.competencia) q.competencia = params.competencia;
  const { data } = await api.get<RoiRanking>("/indicators/roi/operacional", { params: q });
  return data;
}

/** GET /api/v1/indicators/roi/projetos/{project_id} */
export async function fetchProjectRoi(
  projectId: string,
  params?: { competencia?: string; scenario?: string },
): Promise<ProjectRoi> {
  const q: Record<string, string> = { scenario: params?.scenario ?? DEFAULT_SCENARIO_QUERY };
  if (params?.competencia) q.competencia = params.competencia;
  const { data } = await api.get<ProjectRoi>(`/indicators/roi/projetos/${projectId}`, { params: q });
  return data;
}

export interface ConsolidatedRoi {
  competencia: string;
  scenario: string;
  project_ids: string[];
  project_count: number;
  revenue: number;
  cost: number;
  operational_profit: number;
  roi: number | null;
  roi_pct: number | null;
}

/** GET /api/v1/indicators/roi/consolidado (perm: indicators.director) — mês único ou intervalo acumulado. */
export async function fetchConsolidatedRoi(params?: {
  competencia?: string;
  dataInicial?: string;
  dataFinal?: string;
  scenario?: string;
  projectIds?: string[];
}): Promise<ConsolidatedRoi> {
  const q: Record<string, string> = { scenario: params?.scenario ?? DEFAULT_SCENARIO_QUERY };
  if (params?.dataInicial) q.data_inicial = params.dataInicial;
  if (params?.dataFinal) q.data_final = params.dataFinal;
  if (!params?.dataInicial && !params?.dataFinal && params?.competencia) q.competencia = params.competencia;
  if (params?.projectIds && params.projectIds.length > 0) q.project_ids = params.projectIds.join(",");
  const { data } = await api.get<ConsolidatedRoi>("/indicators/roi/consolidado", { params: q });
  return data;
}

export interface RoiEvolutionPoint {
  competencia: string;
  revenue: number;
  cost: number;
  operational_profit: number;
  roi: number | null;
  roi_pct: number | null;
}

export interface RoiEvolution {
  scenario: string;
  project_ids: string[];
  points: RoiEvolutionPoint[];
}

/** GET /api/v1/indicators/roi/evolucao */
export async function fetchRoiEvolution(params: {
  dataInicial: string;
  dataFinal: string;
  scenario?: string;
  projectIds?: string[];
}): Promise<RoiEvolution> {
  const q: Record<string, string> = {
    data_inicial: params.dataInicial,
    data_final: params.dataFinal,
    scenario: params.scenario ?? DEFAULT_SCENARIO_QUERY,
  };
  if (params.projectIds && params.projectIds.length > 0) q.project_ids = params.projectIds.join(",");
  const { data } = await api.get<RoiEvolution>("/indicators/roi/evolucao", { params: q });
  return data;
}

// --- Dashboard Executivo: Evolução Financeira -------------------------------

export interface FinancialEvolutionPoint {
  competencia: string;
  faturamento: number;
  custo_mo: number;
  custo_veiculos: number;
  lucro_operacional: number;
  lucro_liquido: number;
}

export interface FinancialKpi {
  total: number;
  /** crescimento mês inicial→final (%). null = base zero (indefinido). */
  growth_pct: number | null;
}

export interface FinancialKpis {
  faturamento: FinancialKpi;
  custo_mo: FinancialKpi;
  lucro_operacional: FinancialKpi;
  lucro_liquido: FinancialKpi;
}

export interface MonthlyHighlight {
  competencia: string;
  value: number;
}

export interface ProjectHighlight {
  project_id: string;
  project_name: string;
  value: number;
}

export interface FinancialInsights {
  maior_faturamento: MonthlyHighlight | null;
  menor_faturamento: MonthlyHighlight | null;
  maior_lucro_operacional: MonthlyHighlight | null;
  maior_lucro_liquido: MonthlyHighlight | null;
  projeto_maior_faturamento: ProjectHighlight | null;
  projeto_maior_lucro: ProjectHighlight | null;
  tendencia: "alta" | "baixa" | "estavel";
  crescimento_acumulado_pct: number | null;
}

export interface FinancialEvolution {
  scenario: string;
  start: string;
  end: string;
  project_ids: string[];
  cost_centers: string[];
  project_count: number;
  points: FinancialEvolutionPoint[];
  kpis: FinancialKpis;
  insights: FinancialInsights;
}

/** GET /api/v1/indicators/evolucao-financeira — payload agregado único do dashboard executivo. */
export async function fetchFinancialEvolution(params: {
  dataInicial: string;
  dataFinal: string;
  scenario?: string;
  projectIds?: string[];
  costCenters?: string[];
}): Promise<FinancialEvolution> {
  const q: Record<string, string> = {
    data_inicial: params.dataInicial,
    data_final: params.dataFinal,
    scenario: params.scenario ?? DEFAULT_SCENARIO_QUERY,
  };
  if (params.projectIds && params.projectIds.length > 0) q.project_ids = params.projectIds.join(",");
  if (params.costCenters && params.costCenters.length > 0) q.cost_centers = params.costCenters.join(",");
  const { data } = await api.get<FinancialEvolution>("/indicators/evolucao-financeira", { params: q });
  return data;
}

export interface FilterProject {
  project_id: string;
  project_name: string;
  cost_center: string | null;
}

export interface IndicatorFilters {
  projects: FilterProject[];
  cost_centers: string[];
}

/** GET /api/v1/indicators/filtros — opções de filtro (projetos + centros de custo). */
export async function fetchIndicatorFilters(): Promise<IndicatorFilters> {
  const { data } = await api.get<IndicatorFilters>("/indicators/filtros");
  return data;
}

export interface KpiCatalogEntry {
  code: string;
  name: string;
  status: "available" | "coming_soon";
}

/** GET /api/v1/indicators/catalog */
export async function fetchKpiCatalog(): Promise<KpiCatalogEntry[]> {
  const { data } = await api.get<{ items: KpiCatalogEntry[] }>("/indicators/catalog");
  return data.items;
}

File: app/schemas/indicators.py
from __future__ import annotations

from datetime import date
from uuid import UUID

from pydantic import BaseModel


class ProjectRoi(BaseModel):
    """ROI Operacional de um projeto em uma competência.

    ROI = operational_profit / total_cost (do FinancialService).
    `roi` é None quando total_cost <= 0 (investimento indefinido).
    """

    project_id: UUID
    project_name: str
    competencia: date
    scenario: str
    revenue: float | None = None
    cost: float | None = None
    operational_profit: float | None = None
    roi: float | None
    roi_pct: float | None


class RoiRanking(BaseModel):
    """Ranking de ROI Operacional dos projetos ELEGÍVEIS (ordenado desc; None ao fim).

    Elegibilidade = movimentação econômica no período (receita ou custo > 0),
    independente de status; exclui apenas projetos deletados.
    """

    competencia: date
    scenario: str
    # DEPRECADO SEMANTICAMENTE: mantido por compatibilidade de API/frontend.
    # NÃO significa mais "apenas projetos ativos" — agora indica "apenas projetos
    # elegíveis para indicadores" (com movimentação econômica no período; projetos
    # encerrados com receita/custo entram, ativos sem movimentação saem).
    only_active: bool
    items: list[ProjectRoi]


class ConsolidatedRoi(BaseModel):
    """ROI consolidado: Σ operational_profit / Σ total_cost (NUNCA média de ROIs)."""

    competencia: date
    scenario: str
    project_ids: list[UUID]
    project_count: int
    revenue: float | None = None
    cost: float | None = None
    operational_profit: float | None = None
    roi: float | None
    roi_pct: float | None


class RoiEvolutionPoint(BaseModel):
    competencia: date
    revenue: float | None = None
    cost: float | None = None
    operational_profit: float | None = None
    roi: float | None
    roi_pct: float | None


class RoiEvolution(BaseModel):
    scenario: str
    project_ids: list[UUID]
    points: list[RoiEvolutionPoint]


class KpiCatalogEntry(BaseModel):
    code: str
    name: str
    status: str  # "available" | "coming_soon"


class KpiCatalog(BaseModel):
    items: list[KpiCatalogEntry]


# --- Dashboard Executivo: Evolução Financeira -------------------------------


class FinancialEvolutionPoint(BaseModel):
    """Ponto mensal do Dashboard Executivo (sem Combustível, conforme escopo)."""

    competencia: date
    faturamento: float | None = None
    custo_mo: float | None = None
    custo_veiculos: float | None = None
    lucro_operacional: float | None = None
    lucro_liquido: float | None = None


class FinancialKpi(BaseModel):
    """Card de KPI: total acumulado no período + crescimento mês inicial→final."""

    total: float | None = None
    growth_pct: float | None


class FinancialKpis(BaseModel):
    faturamento: FinancialKpi
    custo_mo: FinancialKpi
    lucro_operacional: FinancialKpi
    lucro_liquido: FinancialKpi


class MonthlyHighlight(BaseModel):
    competencia: date
    value: float | None = None


class ProjectHighlight(BaseModel):
    project_id: UUID
    project_name: str
    value: float | None = None


class FinancialInsights(BaseModel):
    """Painel 'Insights' — destaques calculados automaticamente."""

    maior_faturamento: MonthlyHighlight | None
    menor_faturamento: MonthlyHighlight | None
    maior_lucro_operacional: MonthlyHighlight | None
    maior_lucro_liquido: MonthlyHighlight | None
    projeto_maior_faturamento: ProjectHighlight | None
    projeto_maior_lucro: ProjectHighlight | None
    tendencia: str  # "alta" | "baixa" | "estavel"
    crescimento_acumulado_pct: float | None


class FinancialEvolution(BaseModel):
    """Payload completo do Dashboard Executivo de Evolução Financeira."""

    scenario: str
    start: date
    end: date
    project_ids: list[UUID]
    cost_centers: list[str]
    project_count: int
    points: list[FinancialEvolutionPoint]
    kpis: FinancialKpis
    insights: FinancialInsights


class FilterProject(BaseModel):
    project_id: UUID
    project_name: str
    cost_center: str | None


class IndicatorFilters(BaseModel):
    """Opções de filtro disponíveis (apenas dimensões com cadastro estruturado)."""

    projects: list[FilterProject]
    cost_centers: list[str]
